### ISSUE TITLE
optimizer speed for resnet101

### ISSUE THREAD

**Issue by jeageun** [msr-fiddle/pipedream#12]
Hi, I'm trying to compare resnet101 with model parallelism and your pipeline parallelism using a nvprof. 
For this one, I'm trying to make an optimization code to launch. 

I launched the python code below, and it takes a long time, so I can't get results for this code.

`python convert_graph_to_model.py -f resnet101_partitioned/gpus=4.txt -n RESNET101Partitioned -a resnet101 -o ../runtime/image_classification/models/resnet101/gpus=4 --stage_to_num_ranks 0:1,1:1,2:1,3:1
`
How long did you take to create a resnet101 optimization result? Is there a way to get a result faster? If you have already computed the result for this, could you guys upload it for me?

Thanks

**Comment by jeageun** [comment]
I think I find a problem and make improvement in my fork. 
Could you verify it? it gives much faster result for resnet101, but I'm not sure it is really correct. 

https://github.com/jeageun/pipedream/blob/master/graph/graph.py

Thanks

**Comment by deepakn94** [comment]
Can you summarize what the intended purpose of the change is?

**Comment by jeageun** [comment]
I think your implementation label the depth for all layers and your implementation uses the Depth First Search approach. This approach is good if there is no merge point, but if there is merge point, all labeling using a shorter path is not valid. Since you don't remove this invalid path from a work queue, you need to go through the invalid path until it ends. After this invalid path ends, the other valid path always updates depth value.

In my modification code, 
1. I [remove these redundant paths before it ends](https://github.com/jeageun/pipedream/blob/master/graph/graph.py#L112-L118). However, if we use DFS, we can't realize it is valid or not, I use [BFS](https://github.com/jeageun/pipedream/blob/master/graph/graph.py#L101). 
2. I skip building depth if there is the previous result for [depth](https://github.com/jeageun/pipedream/blob/master/graph/graph.py#L91-L93).
